packages/shield-controller/src/polling-with-timeout-abort.ts

```diff
@@ -0,0 +1,200 @@
+export type RequestEntry = {
+  abortController: AbortController; // The abort controller for the request
+  abortHandler: (ev: Event) => void; // The abort handler for the request
+  timerId: NodeJS.Timeout; // The timer ID for the request timeout
+};
+
+export type RequestFn<ReturnType> = (
+  signal: AbortSignal,
+) => Promise<ReturnType>;
+
+export class PollingWithTimeoutAndAbort {
+  readonly ABORT_REASON_TIMEOUT = 'Request timed out';
+
+  readonly ABORT_REASON_CANCELLED = 'Request cancelled';
+
+  // Map of request ID to request entry
+  readonly #requestEntries: Map<string, RequestEntry> = new Map();
+
+  readonly #timeout: number;
+
+  readonly #pollInterval: number;
+
+  constructor(config: { timeout: number; pollInterval: number }) {
+    this.#timeout = config.timeout;
+    this.#pollInterval = config.pollInterval;
+  }
+
+  /**
+   * Poll a request with a timeout and abort.
+   * This will poll the request until it succeeds or fails due to the timeout or the abort signal being triggered.
+   *
+   * @param requestId - The ID of the request to poll.
+   * @param requestFn - The function to poll the request.
+   * @param pollingOptions - The options for the polling.
+   * @param pollingOptions.timeout - The timeout for the request. Defaults to the constructor's timeout.
+   * @param pollingOptions.pollInterval - The interval for the polling. Defaults to the constructor's pollInterval.
+   * @param pollingOptions.fnName - The name of the function to poll the request. Defaults to an empty string.
+   * @returns The result of the request.
+   */
+  async pollRequest<ReturnType>(
+    requestId: string,
+    requestFn: RequestFn<ReturnType>,
+    pollingOptions: {
+      timeout?: number;
+      pollInterval?: number;
+      fnName?: string;
+    } = {},
+  ) {
+    const timeout = pollingOptions.timeout ?? this.#timeout;
+    const pollInterval = pollingOptions.pollInterval ?? this.#pollInterval;
+
+    // clean up the request entry if it exists
+    this.abortPendingRequest(requestId);
```

Comment:
Just curious, why did you rename this from `abortPendingRequests` to `abortPendingRequest`?